RavenDB-7037

IdanHaim opened the pull request:
- DocumentStore now get list Of Urls.
- Wait for UpdateTopology
- New class clusterRequestExecutor for cluster requests
- Fix Tests
- Force UpdateTopology in RequestExecutor

IdanHaim committed:
RavenDB-7037
- DocumentStore now get list Of Urls.
- Wait for UpdateTopology
- New class clusterRequestExecutor for cluster requests
- Fix Tests
- Force UpdateTopology in RequestExecutor

ppekrol commented on src/Raven.Client/Documents/Operations/AdminOperationExecutor.cs:
```diff
@@ -8,26 +8,19 @@
+        private RequestExecutor _requestExecutor;
         private ServerOperationExecutor _serverOperationExecutor;
 
+        private RequestExecutor RequestExecutor => _requestExecutor ?? (_requestExecutor = _store.GetRequestExecutor());
```
I think that you need to pass 'databaseName' to GetRequestExecutor

ayende replied:
Is this for admin on a specific database, or global admin stuff?

ayende replied:
Oh, for a specific db, yes, we need to pass it.

IdanHaim replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/AdminOperationExecutor.cs:
```diff
@@ -8,26 +8,19 @@
+        private RequestExecutor _requestExecutor;
         private ServerOperationExecutor _serverOperationExecutor;
 
+        private RequestExecutor RequestExecutor => _requestExecutor ?? (_requestExecutor = _store.GetRequestExecutor());
```
Private fields should start from _ and lowerCase

ppekrol commented on src/Raven.Client/Http/RequestExecutor.cs:
```diff
@@ -239,44 +168,140 @@ public void Execute<TResult>(RavenCommand<TResult> command, JsonOperationContext
             }
         }
 
+        private bool TryloadFromCache(string url, JsonOperationContext context)
```
we have a typo here TryLoadFromCache

IdanHaim replied:
done

ppekrol commented on src/Raven.Client/Http/RequestExecutor.cs:
```diff
@@ -628,7 +667,7 @@ public class NodeSelector
             public Topology Topology => _topology;
 
-            public int CurrentNodeIndex { get; private set; }
+            private int CurrentNodeIndex;
```
Private fields should start with _ and lowerCase letter

IdanHaim replied:
done

ppekrol commented on src/Raven.Client/Util/ConnectionStringParser.cs:
```diff
@@ -23,19 +41,40 @@ public string Url
+        }
+    }
+
+    public class EmbeddedRavenConnectionStringOptions : RavenConnectionStringOptions
```
This should be deleted. Not needed any more.

IdanHaim replied:
done

ppekrol commented on src/Raven.Client/Util/ConnectionStringParser.cs:
```diff
@@ -23,19 +41,40 @@ public string Url
+        public bool RunInMemory { get; set; }
+    }
+
+    public class FilesConnectionStringOptions : ConnectionStringOptions
```
This should be deleted. Not needed any more.

IdanHaim replied:
done

ppekrol commented on src/Raven.Server/Documents/Handlers/Admin/RachisAdminHandler.cs:
```diff
@@ -76,14 +77,17 @@ public Task GetClusterTopology()
                         ["NodeTag"] = nodeTag
-                    });
+                    };
+                    if (ServerStore.GetClusterErrors().Count > 0)
```
I suggest assigning result from GetClusterErrors to variable to avoid double execution.

ayende replied:
+1, actually need to make sure that we materialize it, not just call it once.

IdanHaim replied:
done

ppekrol commented on src/Raven.Server/Documents/Handlers/Admin/RachisAdminHandler.cs:
```diff
@@ -76,14 +77,17 @@ public Task GetClusterTopology()
-                    });
+                    };
+                    if (ServerStore.GetClusterErrors().Count > 0)
+                        json["Erros"] = ServerStore.GetClusterErrors();
```
We have a typo here: Errors

IdanHaim replied:
done

ppekrol commented on src/Raven.Client/Http/ClusterRequestExecutor.cs:
```diff
@@ -0,0 +1,98 @@
+
+            _lastKnownUrls = initialUrls;
+
+            throw new AggregateException("Failed to retrieve clsuter topology from all known nodes", list);
```
We have a typo here: cluster

ayende replied:
We need to do a match of each node and its error. At least the message of the exception, and then the full details in the `InnerExceptions`

ppekrol commented on src/Raven.Client/Http/RequestExecutor.cs:
```diff
@@ -80,149 +82,76 @@ public class AggressiveCacheOptions
+                    }
+                }
+            });
+            executor.TopologyEtag = -1;
```
Is this preventing topology updates? I cannot find relevant code.

ppekrol commented on src/Raven.Client/Http/RequestExecutor.cs:
```diff
@@ -580,12 +613,17 @@ private void UpdateCurrentTokenCallback(object _)
                 return;
 
             _updateTopologySemaphore.Wait();
+            // _clusterTopologySemaphore.Wait();
```
Can we clean that up?

IdanHaim replied:
done

ppekrol commented on src/Raven.Client/Http/RequestExecutor.cs:
```diff
@@ -595,6 +633,7 @@ public void Dispose()
             _updateCurrentTokenTimer?.Dispose();
             _updateTopologyTimer?.Dispose();
             _updateTopologySemaphore.Dispose();
+            // _clusterTopologySemaphore.Dispose();
```
Can we clean that up?

IdanHaim replied:
done

ppekrol commented on src/Raven.Server/Web/System/DatabasesHandler.cs:
```diff
@@ -79,6 +80,7 @@ public Task GetTopology()
                         HttpContext.Response.StatusCode = (int)HttpStatusCode.ServiceUnavailable;
                         using (var writer = new BlittableJsonTextWriter(context, HttpContext.Response.Body))
                         {
+                            Console.WriteLine("error");
```
Should probably be removed.

IdanHaim replied:
done

ppekrol commented on src/Raven.Server/Web/System/DatabasesHandler.cs:
```diff
@@ -65,9 +65,10 @@ public Task Databases()
         [RavenAction("/topology", "GET", "/topology?name={databaseName:string}&url={url:string}")]
         public Task GetTopology()
         {
-            var name = GetQueryStringValueAndAssertIfSingleAndNotEmpty("name");
```
Not sure if this will not affect the Read method below. Can you recheck that?

IdanHaim replied:
It is not Name is still required and We will get exception.

ayende commented on src/Raven.Client/Documents/DocumentStore.cs:
```diff
@@ -389,8 +391,7 @@ protected override void AfterSessionCreated(InMemoryDocumentSessionOperations se
         {
-            var databaseName = database ?? Database ?? MultiDatabase.GetDatabaseName(Url);
-            var changes = _observeChangesAndEvictItemsFromCacheForDatabases.GetOrDefault(databaseName);
+            var changes = _observeChangesAndEvictItemsFromCacheForDatabases.GetOrDefault(database);
```
I think we need the `database ?? Database` here.

IdanHaim replied:
done

ayende commented on src/Raven.Client/Http/ClusterRequestExecutor.cs:
```diff
@@ -0,0 +1,98 @@
+        {
+        }
+
+        public new static ClusterRequestExecutor Create(string[] urls, string databaseName, string apiKey)
```
Why do we have  database name when we are talking about cluster request manager?

IdanHaim replied:
We need to be able to override those methods. see requestExecutor

ayende replied:
`new static` is a pretty bad idea, better to have a different method name.

ayende replied:
And I'm not following, you can't override a static method.

IdanHaim replied:
done

ayende commented on src/Raven.Client/Http/ClusterRequestExecutor.cs:
```diff
@@ -0,0 +1,98 @@
+            return executor;
+        }
+
+        public new static RequestExecutor CreateForSingleNode(string url, string databaseName, string apiKey)
```
Same

IdanHaim replied:
Same

ayende commented on src/Raven.Client/Http/ClusterRequestExecutor.cs:
```diff
@@ -0,0 +1,98 @@
+            throw new AggregateException("Failed to retrieve clsuter topology from all known nodes", list);
+        }
+
+        public async Task<bool> GetClusterTopologyAsync(ServerNode node, int timeout)
```
Caching is handled by the base class, right? And if so, did you verify that we aren't clashing?

ayende commented on src/Raven.Client/Util/ConnectionStringParser.cs:
```diff
@@ -1,18 +1,36 @@
+            get => _urls;
+            set
+            {
+                if (value == null) throw new ArgumentNullException(nameof(value));
```
Isn't this duplicating the code in the `Store.Urls` ?

IdanHaim replied:
deleted

ayende commented on src/Raven.Client/Util/ConnectionStringParser.cs:
```diff
@@ -93,9 +140,23 @@ protected virtual bool ProcessConnectionStringOption(ConnectionStringOptions opt
-                case "url":
-                    if (string.IsNullOrEmpty(options.Url))
-                        options.Url = value;
+                case "urls":
```
Are there tests for parsing multiple urls?

IdanHaim replied:
http://issues.hibernatingrhinos.com/issue/RavenDB-7239

IdanHaim committed:
RavenDB-7037
- Add ClusterTopology Cache
- fix Typos
- ClusterRequestExecutor no longer need database name
